README
Parallel-DataPipe version 0.10
=================================

CPAN module Parallel::DataPipe for parallel data processing pipe and pipeline

Multiplatform implementation: tests are passed for OpenBSD, MSWin32, Freebsd, Cygwin, Linux


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

It depends only on Core Perl modules now.

if found it uses Sereal module for serialization instead of Storable as the former is more efficient.

SOURCE REPOSITORY

https://github.com/okharch/Parallel-DataPipe

COPYRIGHT AND LICENCE

All right reserved. This library is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.




Changes
Revision history for Perl extension Parallel::DataPipe.
0.10  Thu Jun 20
    - Increased min version of Perl to 5.14.2
0.09  Thu Jun 20
    Merging changes from pull request by Shlomi Fish. Thank you, Shlomi.
    - Add t/pod.t and fix broken POD. (-- Shlomi Fish)
    - Add t/style-trailing-space.t to check for trailing space.

0.08  Thu May 23
    - pipeline() documentation was slightly improved

0.07  Mon May  22
    - pipeline - algorithm of n parallel unix-like pipes implemented, see pipeline() doc

0.06  Mon May  20
    - The interface and behaviour of Parallel::DataPipe::run has been changed.
      For details, see:
      https://github.com/okharch/Parallel-DataPipe/commit/d99a9cb2078e32dfb37d80e24c335f0006184639

0.05  Sun May  19
    - Minor fix for PIPE_MAX_CHUNK_SIZE: be pessimistic unless the platform
      has good news for us.
      16KB for Linux & Cygwin, 1KB for others. This allowed the tests to
      pass on OpenBSD.

0.04  Sat May  18
    - At last! multiplatform implementation. Tests are passing for MSWin32,
      Freebsd, Cygwin, Linux

0.03  Tue May  14 2013
    - Fix for #85289: tests hang on Windows. Added a test to test if
      pipes work as intented [okharch]

0.02  Mon May  13 2013
	- Override SIG{TERM} for child to exit silently from data processors
	- Get rid of single-thread logic
    - Fixup for input_iterator’s returned value, so it can return an
      empty string ('') and 0 for processing. Only undef is considered
      as EOF .
	- $_ is also passed as parameter (@_ or $_[0]) to the process_data closure

0.01  Wed May  8 07:44:03 2013
	- Original version; created by h2xs 1.23 with options
		-AX Parallel::DataPipe


Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;
use IO::Select;
use lib 'lib';
use Parallel::DataPipe;

print "Check if fork works on the platform...\n";
my $fork_ok = eval {
    my $pid = fork();
    if ($pid == 0) {
        exit;
    }
    my $child = wait;
    die "was not able to catch child!" unless $pid == $child;
    1;
};
die "Seems like fork does not work on the platform:\n$@" unless $fork_ok;

print "Check if pipe works here as intented...\n";
eval {
    my ($pipe_read,$pipe_write) = Parallel::DataPipe::pipely();
    my $origin = 'PIPE';
    $pipe_write->syswrite($origin);
    $pipe_read->sysread(my $buf,length($origin));
    die "Pipe read/write failed: content is corrupted" unless $buf eq $origin;
    1;
}  || die "Pipes does not work as intented on this platform, sorry:\n$@";

print "Fork & pipes looks good!\n";

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Parallel::DataPipe',
    VERSION_FROM      => 'lib/Parallel/DataPipe.pm', # finds $VERSION
    PREREQ_PM         => {
        # e.g., Module::Name => 1.1
        #'IO::Pipely' => '0.002',
    }, 
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
     
      (ABSTRACT => 'Parallel Data Processing Conveyor',
       AUTHOR         => 'Oleksandr Kharchenko <okharch@okharch.com>') : ()),
);


lib/Parallel/DataPipe/KISS.pm
# this basic (my favorite keep it simple stupid) implementation
# should work ok for all platforms with fork support
package Parallel::DataPipe::KISS;

use 5.008; # Perl::MinimumVersion says that

use strict;
use warnings;
require Storable;
use File::Slurp qw(read_file write_file);
use File::Temp qw(tempdir);
use Carp qw(confess);

my $parent = $$;

# this should work with Windows NT or if user explicitly set that
my $number_of_cpu_cores = $ENV{NUMBER_OF_PROCESSORS}; 
sub number_of_cpu_cores {
    #$number_of_cpu_cores = $_[0] if @_; # setter
    return $number_of_cpu_cores if $number_of_cpu_cores;
    eval {
        # try unix (linux,cygwin,etc.)
        $number_of_cpu_cores = scalar grep m{^processor\t:\s\d+\s*$},`cat /proc/cpuinfo 2>/dev/null`;
        # try bsd
        ($number_of_cpu_cores) = map m{hw.ncpu:\s+(\d+)},`sysctl -a` unless $number_of_cpu_cores;
    };
    # otherwise it sets number_of_cpu_cores to 2
    return $number_of_cpu_cores || 1;
}

sub store {
	my $self = shift;
	$self->{store}->(@_);
}

sub retrieve {
	my $self = shift;
    confess('no file:'.$_[0]) unless -s $_[0];
	$self->{retrieve}->(@_);
}

# this inits store and retrieve with Storable subroutines
sub _init_serializer {
    my ($self,$param) = @_;
    my ($store, $retrieve) = grep $_ && ref($_) eq 'CODE',map delete $param->{$_},qw(store retrieve);
    if ($store && $retrieve) {
        $self->{store} = $store;
        $self->{retrieve} = $retrieve;
    } else {
        $self->{store} = \&Storable::nstore;
        $self->{retrieve} = \&Storable::retrieve;
    }
}


# this subroutine reads data from pipe and converts it to perl reference
sub _get_data {
    my ($self,$fh) = @_;
    my $data;
    if ($fh =~ m{\.dat$}) {
        $data = read_file $fh, binmode => ':raw';
    } else {
        $data = $self->retrieve($fh);
    }
    unlink ($fh);
    return $data;
}

# this store data to file. serialize if ref
sub _put_data {
    my ($self,$fh,$data) = @_;
    if (ref($data)) {
        $fh .= ".ref";
        $self->store($data,$fh);
    } else {
        $fh .= ".dat";
        write_file $fh, {binmode => ':raw'}, $data;
    }
}

sub _fork_data_processor {
    my ($data_processor_callback) = @_;
    # create processor as fork
    my $pid = fork();
    unless (defined $pid) {
        #print "say goodbye - can't fork!\n"; <>;
        die "can't fork!";
    }
    if ($pid == 0) {
        local $SIG{TERM} = sub {exit;}; # exit silently from data processors
        # data processor is eternal loop which wait for raw data on pipe from main
        # data processor is killed when it's not needed anymore by _kill_data_processors
        $data_processor_callback->() while (1);
        exit;
    }
    return $pid;
}

sub _wait_pipe {
    my ($self,$pipe,@alarm) = @_;
    my @files;
    debug("%s wait for %s",$$,$pipe);
    my $get_pipe = -d $pipe? sub {glob "$pipe/*"} : sub {grep -s, map "$pipe.$_", qw(dat ref) };
    while ( (@files = $get_pipe->()) == 0 ) {
        #_wakeup(@alarm) if @alarm;
        debug("%s wait (loop) for %s",$$,$pipe) unless $parent == $$;        
        eval {
            local $SIG{ALRM} = sub { die "Alarm!\n" };
            sleep;
        };       
    }
    debug("%s caught files %s",$$,\@files);
    wantarray? @files : $files[0];
}

sub _wakeup {
    kill 'SIGALRM', @_;
    #debug('wakeup %d',@_);
}

sub _create_data_processor {
    my ($self,$process_data_callback,$process_num) = @_;
    
    # row data pipe main => processor
    my $raw_data_pipe = sprintf("%s/%04d",$self->{raw_data_dir},$process_num);
    my $processed_data_pipe = sprintf("%s/processed-%04d",$self->{processed_data_dir},$process_num);
    
    my $parent = $$;
 
    my $data_processor = sub {
        my $pipe = $self->_wait_pipe($raw_data_pipe);#,$self->{data_processor_sleep}
        local $_ = $self->_get_data($pipe);
        # process data with given subroutine
        $_ = $process_data_callback->($_);
        # puts processed data back on pipe to main
        $self->_put_data($processed_data_pipe,$_);
        #_wakeup($parent);
    };
    
    # return data processor record 
    return {
        pid => _fork_data_processor($data_processor),  # needed to kill processor when there is no more data to process
        raw_data_pipe => $raw_data_pipe,                 # pipe to write raw data from main to data processor
        #processed_data_pipe => $processed_data_pipe,
    };
}

sub _create_data_processors {
    my ($self,$process_data_callback,$number_of_data_processors) = @_;
    
    $number_of_data_processors = $self->number_of_cpu_cores unless $number_of_data_processors;
    
    die "process_data parameter should be code ref" unless ref($process_data_callback) eq 'CODE';
	confess "\$number_of_data_processors:undefined" unless defined($number_of_data_processors);
    
    return [map $self->_create_data_processor($process_data_callback,$_), 1..$number_of_data_processors];
}

sub process_data {
	my ($self,$data) = @_;

    debug('process data:%s',$data);

    my $free_processor = $self->{free_processors}?
        $self->{free_processors}--
        :
        $self->receive_and_merge_data
    ;

    my $processor = $self->{processors}[$free_processor-1]; # free_processor is 1..n
    $processor->{item_number} = $self->{item_number}++;

    # put raw data to processor's pipe    
    $self->_put_data($processor->{raw_data_pipe},$data);
    
    debug('put item %d to processor %d (%s)',$processor->{item_number},$free_processor,$processor->{raw_data_pipe});

    # let processor know it can fetch the data from pipe
    kill('SIGALRM',$processor->{pid});
}

# this method is called once when input data EOF.
# to find out how many processors were involved in data processing
# then it calls receive_and_merge_data 1..busy_processors times
sub busy_processors {
    my $self = shift;
    my $result = @{$self->{processors}} - $self->{free_processors};
    debug('processors:%d busy:free_processors:%d, busy: %d',scalar(@{$self->{processors}}),$self->{free_processors},$result);
    return $result;
}

sub receive_and_merge_data {
	my $self = shift;
    # look if we have queue of processed items to merge
    my @processed = @{$self->{processed}};
    
    # wait for processed data unless we already have this queue
    my @alarm = map $_->{pid}, @{$self->{processors}};
    @processed = $self->_wait_pipe($self->{processed_data_dir},@alarm) unless @processed;
    
    # take first item from queue
    my $processed = shift(@processed);
    
    # and store the queue for subsequent calls
    $self->{processed} = \@processed;
    
    # read data from processed pipe
    local $_ = $self->_get_data($processed);
    
    # find out how to process it
    my ($processors,$data_merge_code) = @{$self}{qw(processors data_merge_code)};
    
    # find out process_num encoded to filename with data item
    my ($process_num) = $processed =~ m{processed-(\d+)\.};
    
    # this store the item number in source data pipe
    # it's cheap to support and may be someone need to reorder output item in source order
    my $item_number = $processors->[$process_num-1]{item_number};
    
    # merge processed data
    $data_merge_code->($_,$item_number);
    
    # returns the number of processor which is free now to process data
    return $process_num; 
}
    
sub _kill_data_processors {
    my ($processors) = @_;
    my @pid_to_kill = map $_->{pid}, @$processors;
    my %pid_to_wait = map {$_=>undef} @pid_to_kill;
    debug('killing data processors : %s',\@pid_to_kill);
    kill('SIGTERM',@pid_to_kill);
    while (keys %pid_to_wait) {
        my $pid = wait;
        last if $pid == -1;
        delete $pid_to_wait{$pid};
        debug('rip child %s',$pid);
    }
    debug('killed & ripped ok');
}

sub new {
    my ($class, $param) = @_;
	
	my $self = {};
    bless $self,$class;
    # children sleep deply until are waked up by ALRM to read raw pipe
    $self->{data_processor_sleep} = delete $param->{data_processor_sleep} || 999;
    
    # this one sleep less: signal from child destroy the parent
    $self->{main_sleep} = delete $param->{main_sleep} || 999; 
    
    # item_number is number of item from source pipe
    # we support this as a second parameter for data_merge_code subroutine
    # in a case when it wants to reorder output items in source order
    $self->{item_number} = 0;
    
    # queue of procesed but not merged items
    $self->{processed} = [];
    
    # data_merge is sub which merge all processed data inside parent thread
    # it is called each time after process_data returns some new portion of data
    # it has data item both as $_ and $_[0] and also it has item_number as a $_[1]
    $self->{data_merge_code} = delete $param->{'merge_data'};
    die "data_merge should be code ref" unless ref($self->{data_merge_code}) eq 'CODE';
    
    # create bot raw_data_dir & processed_data_dir as a temporary, will remove it in destructor
    $self->{raw_data_dir} = tempdir( CLEANUP => 1 );
    $self->{processed_data_dir} = tempdir( CLEANUP => 1 );
    
    # check if user want to use alternative serialisation routines
    $self->_init_serializer($param);    

    # @$processors is array with data processor info
    $self->{processors} = $self->_create_data_processors(
        map delete $param->{$_},qw(process_data number_of_data_processors)
    );
    
    # this counts processors which are still not involved in processing data
    $self->{free_processors} = @{$self->{processors}};
    debug('free processors:%d',$self->{free_processors});
    
    my $not_supported = join ", ", keys %$param;
    die "Parameters are not supported:". $not_supported if $not_supported;
	
	return $self;
}

sub DESTROY {
	my $self = shift;
    _kill_data_processors($self->{processors});
}

use Data::Dump qw(dump);
sub debug {
	my ($format,@par) = @_;
	my ($package, $filename, $line) = caller;
    if (1) {
        printf STDERR "[%s]%s(%d) $format\n",$$,$filename,$line,map {defined($_)?(ref($_)?dump($_):$_):'undef'} @par if $$==$parent;
    } else {
        open my $fh, ">>","/tmp/$$";
        printf $fh "[%s]%s(%d) $format\n",$$,$filename,$line,map {defined($_)?(ref($_)?dump($_):$_):'undef'} @par if $$==$parent;
        close $fh;
    }
}


1;


lib/Parallel/DataPipe/POSIX.pm
# this basic implementation should work ok for all unix flavors: linux, solaris, bsd, cygwin
package Parallel::DataPipe::POSIX;

use 5.008; # Perl::MinimumVersion says that

use strict;
use warnings;
use Storable;
use IO::Select;
use POSIX;
use constant _EOF_ => (-(2 << 31)+1);
use Carp qw(confess);
#use IPC::SysV qw(IPC_PRIVATE IPC_CREAT IPC_RMID);
use Thread::Semaphore;
# this should work with Windows NT or if user explicitly set that
my $number_of_cpu_cores = $ENV{NUMBER_OF_PROCESSORS}; 
sub number_of_cpu_cores {
    #$number_of_cpu_cores = $_[0] if @_; # setter
    return $number_of_cpu_cores if $number_of_cpu_cores;
    eval {
        # try unix (linux,cygwin,etc.)
        $number_of_cpu_cores = scalar grep m{^processor\t:\s\d+\s*$},`cat /proc/cpuinfo 2>/dev/null`;
        # try bsd
        ($number_of_cpu_cores) = map m{hw.ncpu:\s+(\d+)},`sysctl -a` unless $number_of_cpu_cores;
    };
    # otherwise it sets number_of_cpu_cores to 2
    return $number_of_cpu_cores || 1;
}

sub freeze {
	my $self = shift;
	$self->{freeze}->(@_);
}

sub thaw {
	my $self = shift;
	$self->{thaw}->(@_);
}

# this inits freeze and thaw with Storable subroutines and try to replace them with Sereal counterparts
sub _init_serializer {
    my ($self,$param) = @_;
    my ($freeze,$thaw) = grep $_ && ref($_) eq 'CODE',map delete $param->{$_},qw(freeze thaw);
    if ($freeze && $thaw) {
        $self->{freeze} = $freeze;
        $self->{thaw} = $thaw;
    } else {
        $self->{freeze} = \&Storable::nfreeze;
        $self->{thaw} = \&Storable::thaw;
        return;
        # try cereal        
        eval q{
            use Sereal qw(encode_sereal decode_sereal);
            $self->{freeze} = \&encode_sereal;
            $self->{thaw} = \&decode_sereal;
        };
    }
    # don't make any assumptions on serializer capabilities, give all the power to user ;)
    # die "bad serializer!" unless join(",",@{$thaw->($freeze->([1,2,3]))}) eq '1,2,3';
}


# this subroutine reads data from pipe and converts it to perl reference
# or scalar - if size is negative
# it always expects size of frozen scalar so it know how many it should read
# to feed thaw 
sub _get_data {
    my ($self,$fh) = @_;
    my $get_process_num = wantarray; # this means it expects process_num
    my ($data_size,$data,$process_num);
    if ($get_process_num) {
        POSIX::read($fh,$process_num,2);
        $process_num = unpack("S*",$process_num);
    }
    POSIX::read($fh,$data_size,4);
    $data_size = unpack("l",$data_size);
    if ($data_size != _EOF_) {
        if ($data_size == 0) {
            $data = '';
        } else {
            POSIX::read($fh,$data,abs($data_size));
            $data = $self->thaw($data) if $data_size>0;            
        }
    }
    debug("got data from %s : %s",($get_process_num?"child $process_num":"parent"),$data);
    return $get_process_num?($process_num,$data):$data;
}

# this subroutine serialize data reference. otherwise 
# it puts negative size of scalar and scalar itself to pipe.
# parameter $process_num defined means it's child who is writing
# to shared pipe to communicate with parent
sub _put_data {
    my ($self,$fh,$data,$process_num) = @_;
    if (defined($process_num)) {
        # pipe is shared, must use semaphore
        debug("child %s semaphore entering %s",$process_num,$data);
        semaphore_up($self->{sem_id});
        debug("child %s semaphore protected %s",$process_num,$data);
        POSIX::write($fh,pack("S*",$process_num),2);
    }
    debug("put data from %s : %s",(defined($process_num)?"child $process_num":"parent"),$data);
    if (!defined($data)) {
        POSIX::write($fh,pack("l",_EOF_),4);
    } elsif (ref($data)) {
        $data = $self->freeze($data);
        # length of serialized data
        POSIX::write($fh, pack("l", length($data)), 4); 
        POSIX::write($fh, $data, length($data));
    } else {
        # (negative) length of scalar
        POSIX::write($fh, pack("l", -length($data)), 4);
        POSIX::write($fh, $data, length($data));
    }
    if (defined($process_num)) {
        debug("child %s semaphore freeing %s",$process_num,$data);
        semaphore_down($self->{sem_id});
    }
}

sub _fork_data_processor {
    my ($data_processor_callback) = @_;
    # create processor as fork
    local $SIG{TERM} = sub {exit;}; # exit silently from data processors
    my $pid = fork();
    unless (defined $pid) {
        #print "say goodbye - can't fork!\n"; <>;
        die "can't fork!";
    }
    if ($pid == 0) {
        # data processor is eternal loop which wait for raw data on pipe from main
        # data processor is killed when it's not needed anymore by _kill_data_processors
        $data_processor_callback->() while (1);
        exit;
    }
    return $pid;
}

sub _create_data_processor {
    my ($self,$process_data_callback,$process_num) = @_;
    
    # row data pipe main => processor
    my ($read, $write) = POSIX::pipe();
 
    my $data_processor = sub {
	# wait for data from raw data pipe
        local $_ = $self->_get_data($read);
        # process data with given subroutine
        $_ = $process_data_callback->($_);
        # puts processed data back on pipe to main
        $self->_put_data($self->{write_pipe},$_,$process_num);
    };
    
    # return data processor record 
    return {
        pid => _fork_data_processor($data_processor),  # needed to kill processor when there is no more data to process
        write_raw_data_pipe => $write,                 # pipe to write raw data from main to data processor 
    };    
}

sub _create_data_processors {
    my ($self,$process_data_callback,$number_of_data_processors) = @_;
    $number_of_data_processors = $self->number_of_cpu_cores unless $number_of_data_processors;
    die "process_data parameter should be code ref" unless ref($process_data_callback) eq 'CODE';
	confess "\$number_of_data_processors:undefined" unless defined($number_of_data_processors);
    return [map $self->_create_data_processor($process_data_callback,$_), 0..$number_of_data_processors-1];
}

sub process_data {
	my ($self,$data) = @_;
    debug('process data:%s',$data);
    my $free_processors = $self->{free_processors};
    my $free_processor;
    if ($free_processors) {
        $free_processor = --$self->{free_processors};
    } else {
        # wait for some which processed date
        $free_processor = $self->receive_and_merge_data;
    }
    my $processor = $self->{processors}[$free_processor];
    $processor->{item_number} = $self->{item_number}++;
    $self->_put_data($processor->{write_raw_data_pipe},$data);
}

# this method is called once when input data EOF.
# to find out how many processors were involved in data processing
# then it calls receive_and_merge_data 1..busy_processors times
sub busy_processors {
    my $self = shift;
    return @{$self->{processors}} - $self->{free_processors};
}

sub receive_and_merge_data {
	my $self = shift;
    my $read_pipe = $self->{read_pipe};
    my ($processors,$data_merge_code) = @{$self}{qw(processors data_merge_code)};
    my $process_num;
    local $_;
    ($process_num,$_) = $self->_get_data($read_pipe);
    my $item_number = $processors->[$process_num]{item_number};
    $data_merge_code->($_,$item_number);
    return $process_num;
}
    
sub _kill_data_processors {
    my ($processors) = @_;
    my @pid_to_kill = map $_->{pid}, @$processors;
    my %pid_to_wait = map {$_=>undef} @pid_to_kill;
    debug('killing data processors : %s',\@pid_to_kill);
    kill('SIGTERM',@pid_to_kill);
    while (keys %pid_to_wait) {
        my $pid = wait;
        last if $pid == -1;
        delete $pid_to_wait{$pid};
        debug('rip child %s',$pid);
    }
    debug('killed & ripped ok');
}

sub new {
    my ($class, $param) = @_;
	
	my $self = {};
    bless $self,$class;
    
    # merge item_number implementation
    $self->{item_number} = 0;
    
    # shared parent pipe. every child uses it to write processed results back to parent
    my ($read, $write) = POSIX::pipe();
    $self->{read_pipe} = $read;
    $self->{write_pipe} = $write;
    
    # semaphore to sync writing to parent pipe
    #$self->{sem_id} = semget(IPC_PRIVATE, 1, 0666 | IPC_CREAT);
    $self->{sem_id} = Thread::Semaphore->new();
	
    # check if user want to use alternative serialisation routines
    $self->_init_serializer($param);    

    # @$processors is array with data processor info
    $self->{processors} = $self->_create_data_processors(
        map delete $param->{$_},qw(process_data number_of_data_processors)
    );
    # this counts processors which are still not involved in processing data
    $self->{free_processors} = @{$self->{processors}};
    debug('free processors:%d',$self->{free_processors});
    
    # data_merge is sub which merge all processed data inside parent thread
    # it is called each time after process_data returns some new portion of data    
    $self->{data_merge_code} = delete $param->{'merge_data'};
    die "data_merge should be code ref" unless ref($self->{data_merge_code}) eq 'CODE';
    
    my $not_supported = join ", ", keys %$param;
    die "Parameters are not supported:". $not_supported if $not_supported;
	
	return $self;
}

sub DESTROY {
	my $self = shift;
    _kill_data_processors($self->{processors});
    #semctl($self->{sem_id},0,IPC_RMID,0);
}

sub semaphore_up { 
    my ($sem_id) = @_;
    $sem_id->up; return;
    my $sem_op = pack("s!*",
        0, 0, 0, # wait until zero
        0, 1, 0, # up
    );
    semop($sem_id, $sem_op)   || die "semop: $!";
}

sub semaphore_down {
    my ($sem_id) = @_;
    $sem_id->down; return;
    my $sem_op = pack("s!*",
        0, -1, 0, # down
    );
    semop($sem_id, $sem_op)   || die "semop: $!";
}


use Data::Dump qw(dump);
sub debug {
    return;
	my ($format,@par) = @_;
    return unless $format =~ m{sema};
	my ($package, $filename, $line) = caller;
	printf STDERR "%s(%d) $format\n",$filename,$line,map {defined($_)?(ref($_)?dump($_):$_):'undef'} @par;
}


1;


lib/Parallel/DataPipe/Conveyor.pm
# this basic implementation should work ok for all unix flavors: linux, solaris, bsd, cygwin
package Parallel::DataPipe::Conveyor;

our $VERSION='0.03';
use 5.008; # Perl::MinimumVersion says that

use strict;
use warnings;
use Storable qw(freeze thaw);
use IO::Select;
use POSIX ":sys_wait_h";
use constant _EOF_ => (-(2 << 31)+1);
use Carp qw(confess);

# this should work with Windows NT or if user explicitly set that
my $number_of_cpu_cores = $ENV{NUMBER_OF_PROCESSORS}; 
sub number_of_cpu_cores {
    #$number_of_cpu_cores = $_[0] if @_; # setter
    return $number_of_cpu_cores if $number_of_cpu_cores;
    eval {
        # try unix (linux,cygwin,etc.)
        $number_of_cpu_cores = scalar grep m{^processor\t:\s\d+\s*$},`cat /proc/cpuinfo 2>/dev/null`;
        # try bsd
        ($number_of_cpu_cores) = map m{hw.ncpu:\s+(\d+)},`sysctl -a` unless $number_of_cpu_cores;
    };
    # otherwise it sets number_of_cpu_cores to 2
    return $number_of_cpu_cores || 1;
}

sub freeze {
	my $self = shift;
	$self->{freeze}->(@_);
}

sub thaw {
	my $self = shift;
	$self->{thaw}->(@_);
}

# this inits freeze and thaw with Storable subroutines and try to replace them with Sereal counterparts
sub _init_serializer {
    my ($self,$param) = @_;
    if (exists $param->{freeze} && ref($param->{freeze}) eq 'CODE' &&  exists $param->{thaw} && ref($param->{thaw}) eq 'CODE') {
        $self->{freeze} = $param->{freeze};
        $self->{thaw} = $param->{thaw};
    } else {
        $self->{freeze} = \&freeze;
        $self->{thaw} = \&thaw;
        # try cereal 
        eval q{
            use Sereal qw(encode_sereal decode_sereal);
            $self->{freeze} = \&encode_sereal;
            $self->{thaw} = \&decode_sereal;
        };
    }
    # don't make any assumptions on serializer capabilities, give all the power to user ;)
    # die "bad serializer!" unless join(",",@{$thaw->($freeze->([1,2,3]))}) eq '1,2,3';
}


# this subroutine reads data from pipe and converts it to perl reference
# or scalar - if size is negative
# it always expects size of frozen scalar so it know how many it should read
# to feed thaw 
sub _get_data { my ($self,$fh) = @_;    
    my ($data_size,$data);
    read $fh,$data_size,4;
    $data_size = unpack("l",$data_size);
    return undef if ($data_size == _EOF_);
    return  "" if $data_size == 0;
    read $fh,$data,abs($data_size);
    $data = $self->{thaw}->($data) if $data_size>0;
    return $data;
}

# this subroutine serialize data reference. otherwise 
# it puts negative size of scalar and scalar itself to pipe.
# it also makes flush to avoid buffering blocks execution
sub _put_data { my ($self,$fh,$data) = @_;
    if (!defined($data)) {
        print $fh pack("l",_EOF_);
    } elsif (ref($data)) {
        $data = $self->{freeze}->($data);
        print $fh pack("l", length($data)).$data;
    } else {
        print $fh pack("l",-length($data)).$data;        
    }
    $fh->flush();
}

sub _fork_data_processor {
    my ($data_processor_callback) = @_;
    # create processor as fork
    local $SIG{TERM} = sub {exit;}; # exit silently from data processors
    my $pid = fork();
    unless (defined $pid) {
        #print "say goodbye - can't fork!\n"; <>;
        die "can't fork!";
    }
    if ($pid == 0) {
        # data processor is eternal loop which wait for raw data on pipe from main
        # data processor is killed when it's not needed anymore by _kill_data_processors
        $data_processor_callback->() while (1);
        exit;
    }
    return $pid;
}

sub _create_data_processor {
    my ($self,$process_data_callback) = @_;
    # row data pipe main => processor
    pipe(my $read_raw_data_pipe,my $write_raw_data_pipe);
    
    # processed data pipe processor => main
    pipe(my $read_processed_data_pipe,my $write_processed_data_pipe);
    
    my $data_processor = sub {
	# wait for data from raw data pipe
        local $_ = $self->_get_data($read_raw_data_pipe);
        # process data with given subroutine
        $_ = $process_data_callback->($_);
        # puts processed data back on pipe to main
        $self->_put_data($write_processed_data_pipe,$_);
    };
    
    # return data processor record 
    return {
        pid => _fork_data_processor($data_processor),            # needed to kill processor when there is no more data to process
        write_raw_data_pipe => $write_raw_data_pipe,            # pipe to write raw data from main to data processor 
        read_processed_data_pipe => $read_processed_data_pipe,  # pipe to read processed data from processor to main thread                                                                    
        is_free => 1,                                           # flag whether processor is free for processing data
                                                                # (waits for data on read_raw_data_pipe )
    };    
}

sub _create_data_processors {
    my ($self,$process_data_callback,$number_of_data_processors) = @_;
    die "process_data parameter should be code ref" unless ref($process_data_callback) eq 'CODE';
	confess "\$number_of_data_processors:undefined" unless defined($number_of_data_processors);
    return [map $self->_create_data_processor($process_data_callback), 1..$number_of_data_processors];
}

sub process_data {
	my ($self,$data) = @_;
    my $processors = $self->{processors};
    my @free_processors = grep $_->{is_free},@$processors;
    return 1 unless @free_processors;
    my $processor = shift(@free_processors);
    $self->_put_data($processor->{write_raw_data_pipe},$data);
    $processor->{is_free} = 0; # now it's busy
    return 0; 
}

sub receive_and_merge_data {
	my $self = shift;
    my ($processors,$data_merge_code) = @{$self}{qw(processors data_merge_code)};
    
    my @busy_processors = grep $_->{is_free}==0,@$processors;
    
    # returns false if there is no busy processors
    return 0 unless @busy_processors;
    
    # blocking read until at least one handle is ready
    my @read_processed_data_pipe = IO::Select->new(map $_->{read_processed_data_pipe},@busy_processors)->can_read();
    for my $rh (@read_processed_data_pipe) {
        local $_ = $self->_get_data($rh);
        $data_merge_code->($_);
        $_->{is_free} = 1 for grep $_->{read_processed_data_pipe} == $rh, @busy_processors;
    }
    
    # returns true if there was busy processors 
    return 1;        
}
    
sub _kill_data_processors {
    my ($processors) = @_;
    my @pid_to_kill = map $_->{pid}, @$processors;
    my %pid_to_wait = map {$_=>undef} @pid_to_kill;
    kill('SIGTERM',@pid_to_kill);
    while (keys %pid_to_wait) {
        my $pid = wait;
        last if $pid == -1;
        delete $pid_to_wait{$pid};
    }
}

sub new {
    my ($class, $param) = @_;
	
	my $self = {};
    bless $self,$class;
	
    # check if user want to use alternative serialisation routines
    $self->_init_serializer($param);    

    # @$processors is array with data processor info
    $self->{processors} = $self->_create_data_processors($param->{'process_data'},$param->{'number_of_data_processors'} || $class->number_of_cpu_cores); 
    
    # data_merge is sub which merge all processed data inside parent thread
    # it is called each time after process_data returns some new portion of data    
    $self->{data_merge_code} = $param->{'merge_data'};
    die "data_merge should be code ref" unless ref($self->{data_merge_code}) eq 'CODE';
	bless $self, $class;
	return $self;
}

sub DESTROY {
	my $self = shift;
    _kill_data_processors($self->{processors});	
}

1;
